### DISCUSSION TITLE
Periodic function execution?

### DISCUSSION THREAD

**Discussion by epsilon-0** [uNetworking/uWebSockets#1332] [Q&A] [answered]
I have a use case where I want to run a function (to clean up things) every X seconds.
Currently, I don't see how to do this with the App.h or Loop.h (maybe I missed some thing?).

Is there any way to periodically run a function here?

Thanks!

**Comment by ghost** [comment]
A timer?

**Reply by epsilon-0** [reply]
Are these in uWebSockets? or should I look at usockets for a lower level of access?

**Comment by Jacob-Burckhardt** [comment, marked as answer]
This calls send every 1000 milliseconds and follows the requirements of the library for thread-safe code:

```
#include "App.h"
   
using namespace std;

struct PerSocketData {};

uWS::WebSocket<false, true, PerSocketData> *gws=nullptr;
    
int main() {
   auto loop = uWS::Loop::get();

   struct us_timer_t *delayTimer = us_create_timer((struct us_loop_t *) loop, 0, 0);

   us_timer_set(delayTimer, [](struct us_timer_t *) {
                               if (gws) {
                                  cout << "calling send" << endl;
                                  gws->send("from server", uWS::OpCode::TEXT);
                               }
                            }, 1000, 1000);
   
   uWS::App app;

   app.ws<PerSocketData>("/*", {
         .idleTimeout = 0,
         .sendPingsAutomatically = false,
         .open = [](auto *ws) {
                    gws = ws;
                 },
         .close = [](auto */*ws*/, int /*code*/, std::string_view /*message*/) {
                     gws = nullptr;
                  }
      }).listen(9001, [](auto *) {
                      });

   app.run();
}
```

**Reply by epsilon-0** [reply]
Thanks!

**Reply by epsilon-0** [reply]
!!!! This worked nicely, thanks!!

**Reply by alexxmanea** [reply]
Sorry for waking up this post. Is it possible to pass the "app", instead of "gws", to the timer's callback function? I want to make a global publish, to every socket connected.

**Reply by ghost** [reply]
Easiest to make the app global

**Reply by alexxmanea** [reply]
I understood that this should "follow the requirements of the library for thread-safe code", because it runs single threaded, on the same thread as the App. But will this also work fine in a multi-threaded uWS::App?

**Reply by ghost** [reply]
Make it global but thread local?
